src/mon/MDSMonitor.cc

```diff
@@ -1680,6 +1694,80 @@ void MDSMonitor::check_sub(Subscription *sub)
   }
 }
 
+void MDSMonitor::update_mds_metadata(mds_gid_t gid,
+				     const map<string, string>& metadata)
+{
+  if (metadata.empty()) {
+    return;
+  }
+  bufferlist bl;
+  int err = mon->store->get(MDS_METADATA_PREFIX, "last_metadata", bl);
+  map<mds_gid_t, Metadata> last_metadata;
+  if (!err) {
+    bufferlist::iterator iter = bl.begin();
+    ::decode(last_metadata, iter);
+  }
+  last_metadata[gid] = metadata;
+
+  MonitorDBStore::TransactionRef t = paxos->get_pending_transaction();
+  bl.clear();
+  ::encode(last_metadata, bl);
+  t->put(MDS_METADATA_PREFIX, "last_metadata", bl);
+  paxos->trigger_propose();
+}
+
+int MDSMonitor::load_metadata(map<mds_gid_t, Metadata>& m)
+{
+  bufferlist bl;
+  int r = mon->store->get(MDS_METADATA_PREFIX, "last_metadata", bl);
+  if (r)
+    return r;
+  bufferlist::iterator it = bl.begin();
+  ::decode(m, it);
+  return 0;
+}
+
+int MDSMonitor::get_mds_metadata(mds_gid_t gid, Formatter *f, ostream& err)
+{
+  assert(f);
+  map<mds_gid_t, Metadata> metadata;
+
+  if (int r = load_metadata(metadata)) {
+    return r;
+  }
+  if (!metadata.count(gid)) {
+    return -EINVAL;
+  }
+  const Metadata& m = metadata[gid];
+  for (Metadata::const_iterator p = m.begin(); p != m.end(); ++p) {
+    f->dump_string(p->first.c_str(), p->second);
+  }
+  return 0;
+}
+
+int MDSMonitor::print_nodes(Formatter *f)
+{
+  assert(f);
+
+  f->open_object_section("mds");
```

Comment:
You're outputting an object with N keys, where you should be outputting an array of N objects with one key each, or an array of N strings.  For example, on my system (with many services on the same hostname):
"""
    {
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain",
        "hostname": "schist.localdomain"
    },
"""

That's actually unreadable JSON because when we load it all the "hostname" keys collapse into one.